keps/sig-api-machinery/20190206-watch-bookmark.md

```diff
@@ -0,0 +1,184 @@
+  - sig-scalability
+reviewers:
+  - "@jpbetz"
+approvers:
+  - "@deads2k"
+  - "@lavalamp"
+creation-date: 2019-02-06
+status: implementable
+see-also:
+  - https://github.com/kubernetes/kubernetes/issues/73585
+replaces:
+  - n/a
+superseded-by:
+  - n/a
+---
+
+# Watch bookmark
+
+## Table of Contents
+
+Table of Contents
+=================
+
+* [Watch bookmark](#watch-bookmark)
+   * [Table of Contents](#table-of-contents)
+   * [Summary](#summary)
+   * [Motivation](#motivation)
+      * [Goals](#goals)
+      * [Non-Goals](#non-goals)
+   * [Proposal](#proposal)
+      * [Risks and Mitigations](#risks-and-mitigations)
+   * [Graduation Criteria](#graduation-criteria)
+   * [Implementation History](#implementation-history)
+   * [Rejected alternatives](#rejected-alternatives)
+      * [Cache in kube-apiserver](#cache-in-kube-apiserver)
+      * [API for send bookmark](#api-for-send-bookmark)
+
+## Summary
+
+Watch API is one of the fundaments of Kubernetes API. The recommended pattern
+for using watch API is to retrieve a collection of resources using consistent
+list and then initiate a watch starting from a resourceVersion returned by the
+list operation. If the client watch is disconnected, a new one can be restarted
+from the last returned resourceVersion.
+
+This proposal make restarting watches cheaper from kube-apiserver performance
+perspective.
+
+## Motivation
+
+While running different scalability tests we observed that restarting watches
+may cause significant load on kube-apiserver when watcher is observing a small
+percentage of changes (due to field or label selector). In extreme cases,
+reestablishing such watcher may even lead to falling out of history window
+and "resource version too old" errors (that requires full relist for that
+watcher).
+
+The reason for that is the fact that even if the last item received by watcher
+has resourceVersion rv1, we may already know that there aren't any changes
+a given watcher is interested in up to rv2 (rv1 < rv2), but we don't have any
+way of communicating it to the watcher. As a result, when restarting a watch,
+client again sends rv1 as a starting point, and we process all events with
+resourceVersion between rv1 and rv2 again unnecessarily.
+
+The proposal presents a proper solution for that problem.
+
+### Goals
+
+- Reduce load on apiserver by minimizing amount of unnecessary watch events
+that need to be processed after restarting a watch.
+- Reduce amount of undesired "resource version too old" errors on reestablishing
+a watch.
+
+### Non-Goals
+
+The following are nice-to-haves, but not primary goals:
+
+- Improve overall watch throughput and/or latency.
+
+## Proposal
+
+We propose introducing a new type of watch event called `Bookmark`. With that
+change, the possible watch event types will be:
+```
+  Added    EventType = "ADDED"
+  Modified EventType = "MODIFIED"
+  Deleted  EventType = "DELETED"
+  Error    EventType = "ERROR"
+  Bookmark EventType = "BOOKMARK"
```

Comment:
Bookmark sounds reasonable. The other names that came to mind were "watermark" and "progress notification".